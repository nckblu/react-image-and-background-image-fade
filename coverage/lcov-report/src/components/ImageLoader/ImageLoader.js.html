<!doctype html>
<html lang="en">
<head>
    <title>Code coverage report for src/components/ImageLoader/ImageLoader.js</title>
    <meta charset="utf-8" />
    <link rel="stylesheet" href="../../../prettify.css" />
    <link rel="stylesheet" href="../../../base.css" />
    <meta name="viewport" content="width=device-width, initial-scale=1">
    <style type='text/css'>
        .coverage-summary .sorter {
            background-image: url(../../../sort-arrow-sprite.png);
        }
    </style>
</head>
<body>
<div class='wrapper'>
  <div class='pad1'>
    <h1>
      <a href="../../../index.html">All files</a> / <a href="index.html">src/components/ImageLoader</a> ImageLoader.js
    </h1>
    <div class='clearfix'>
      <div class='fl pad1y space-right2'>
        <span class="strong">78.26% </span>
        <span class="quiet">Statements</span>
        <span class='fraction'>18/23</span>
      </div>
      <div class='fl pad1y space-right2'>
        <span class="strong">50% </span>
        <span class="quiet">Branches</span>
        <span class='fraction'>5/10</span>
      </div>
      <div class='fl pad1y space-right2'>
        <span class="strong">88.89% </span>
        <span class="quiet">Functions</span>
        <span class='fraction'>8/9</span>
      </div>
      <div class='fl pad1y space-right2'>
        <span class="strong">81.82% </span>
        <span class="quiet">Lines</span>
        <span class='fraction'>18/22</span>
      </div>
    </div>
  </div>
  <div class='status-line medium'></div>
<pre><table class="coverage">
<tr><td class="line-count quiet">1
2
3
4
5
6
7
8
9
10
11
12
13
14
15
16
17
18
19
20
21
22
23
24
25
26
27
28
29
30
31
32
33
34
35
36
37
38
39
40
41
42
43
44
45
46
47
48
49
50
51
52
53
54
55
56
57
58
59
60
61
62
63
64
65
66
67
68
69
70
71
72
73
74
75
76
77
78
79
80
81
82
83
84
85
86
87
88
89
90
91
92
93
94
95
96
97
98
99
100
101
102
103
104
105
106
107
108
109
110
111
112
113
114
115
116
117
118
119
120
121
122
123
124</td><td class="line-coverage quiet"><span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-yes">21x</span>
<span class="cline-any cline-yes">21x</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-yes">26x</span>
<span class="cline-any cline-yes">1x</span>
<span class="cline-any cline-yes">1x</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-no">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-yes">47x</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-yes">47x</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-yes">47x</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-yes">47x</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-yes">22x</span>
<span class="cline-any cline-yes">22x</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-yes">21x</span>
<span class="cline-any cline-yes">21x</span>
<span class="cline-any cline-yes">21x</span>
<span class="cline-any cline-yes">21x</span>
<span class="cline-any cline-yes">3x</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-yes">1x</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-no">&nbsp;</span>
<span class="cline-any cline-no">&nbsp;</span>
<span class="cline-any cline-no">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-yes">3x</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span></td><td class="text"><pre class="prettyprint lang-js">import React from "react";
import PropTypes from "prop-types";
import preloader from "image-preloader";
import { cssTimeToMs } from "../../util";
import defaults from "../../defaults";
import VisibilitySensor from "react-visibility-sensor";
&nbsp;
export class ImageLoader extends React.Component {
  static propTypes = {
    src: PropTypes.string.isRequired,
    transitionTime: PropTypes.string,
    lazyLoad: PropTypes.bool,
    children: PropTypes.oneOfType([
      PropTypes.arrayOf(PropTypes.node),
      PropTypes.node,
      PropTypes.func,
    ]),
  };
&nbsp;
  state = {
    hasLoaded: false,
    hasFailed: false,
    shouldShowLoader: true,
    visibilitySensorIsActive: true,
  };
&nbsp;
  componentDidMount() {
    <span class="missing-if-branch" title="else path not taken" >E</span>if (!this.props.lazyLoad) {
      this.preloadImage();
    }
  }
&nbsp;
  componentDidUpdate(prevProps) {
    if (prevProps.src !== this.props.src) {
      <span class="missing-if-branch" title="else path not taken" >E</span>if (!this.props.lazyLoad) {
        this.preloadImage();
      } else {
<span class="cstat-no" title="statement not covered" >        this.setState({ visibilitySensorIsActive: true });</span>
      }
    }
  }
&nbsp;
  render() {
    const { src, children, lazyLoad } = this.props;
    const {
      hasLoaded,
      hasFailed,
      shouldShowLoader,
      visibilitySensorIsActive,
    } = this.state;
    /*
     * When using ImageLoader the children prop should be a function.
     * Render calls this function with the props it requires to
     * fade in an image.
     * See https://reactjs.org/docs/render-props.html
     */
    const childElement = children({
      hasLoaded,
      shouldShowLoader,
      hasFailed,
      src,
    });
&nbsp;
    /*
     * Wrap element in react visibility sensor when lazyLoad is true
     */
    return lazyLoad ? (
<span class="branch-0 cbranch-no" title="branch not covered" >      &lt;VisibilitySensor</span>
        active={visibilitySensorIsActive}
        onChange={this.handleVisibilityChange}
        partialVisibility
      &gt;
        {childElement}
      &lt;/VisibilitySensor&gt;
    ) : (
      childElement
    );
  }
&nbsp;
  preloadImage() {
    const { src } = this.props;
    preloader
      .simplePreload(src)
      .then(this.handleImageLoaderLoaded)
      .catch(this.handleImageLoaderFailed);
  }
&nbsp;
  /*
   * Once the image is loaded successfully this handler will be
   * called, this will change the `hasLoaded` attribute to true
   * then the `shouldShowLoader` attribute to false after
   * the transition has completed. The reason we do this on a
   * timeout is to ensure the loader is not hidden as soon as the
   * image has loaded resulting in a 'flash'
   */
  handleImageLoaderLoaded = () =&gt; {
    const { transitionTime } = this.props;
    const hideLoaderDelay = cssTimeToMs(transitionTime);
    this.setState({ hasLoaded: true }, () =&gt; {
      setTimeout(() =&gt; {
        this.setState({
          shouldShowLoader: false,
        });
      }, hideLoaderDelay);
    });
  };
&nbsp;
  handleImageLoaderFailed = () =&gt; {
    this.setState({ hasFailed: true });
  };
&nbsp;
  handleVisibilityChange = <span class="fstat-no" title="function not covered" >is</span>Visible =&gt; {
<span class="cstat-no" title="statement not covered" >    if (!isVisible) <span class="cstat-no" title="statement not covered" >return;</span></span>
<span class="cstat-no" title="statement not covered" >    this.setState({ visibilitySensorIsActive: false });</span>
<span class="cstat-no" title="statement not covered" >    this.preloadImage();</span>
  };
}
&nbsp;
ImageLoader.defaultProps = {
  transitionTime: defaults.transitionTime,
};
&nbsp;
export default ImageLoader;
&nbsp;</pre></td></tr>
</table></pre>
<div class='push'></div><!-- for sticky footer -->
</div><!-- /wrapper -->
<div class='footer quiet pad2 space-top1 center small'>
  Code coverage
  generated by <a href="https://istanbul.js.org/" target="_blank">istanbul</a> at Wed Feb 13 2019 18:14:31 GMT+0000 (GMT)
</div>
</div>
<script src="../../../prettify.js"></script>
<script>
window.onload = function () {
        if (typeof prettyPrint === 'function') {
            prettyPrint();
        }
};
</script>
<script src="../../../sorter.js"></script>
</body>
</html>
