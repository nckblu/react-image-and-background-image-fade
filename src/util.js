export const cssTimeToMs = cssTime => {
  const num = parseFloat(cssTime, 10);
  let unit = cssTime.match(/m?s/);
  let milliseconds;

  if (unit) {
    unit = unit[0];
  }

  switch (unit) {
    case "s":
      milliseconds = num * 1000;
      break;
    case "ms":
      milliseconds = num;
      break;
    default:
      milliseconds = 0;
      break;
  }

  return milliseconds;
};

export const getAspectRatioPercentage = (width, height) => {
  const widthNumber = parseInt(width, 10);
  const heightNumber = parseInt(height, 10);
  const ratio = (heightNumber / widthNumber) * 100;
  return `${ratio}%`;
};
