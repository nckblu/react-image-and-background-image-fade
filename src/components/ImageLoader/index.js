import ImageLoader from "./ImageLoader";

export default ImageLoader;
