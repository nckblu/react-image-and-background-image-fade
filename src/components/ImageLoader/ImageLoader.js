import React from "react";
import PropTypes from "prop-types";
import preloader from "image-preloader";
import { cssTimeToMs } from "../../util";
import defaults from "../../defaults";
import VisibilitySensor from "react-visibility-sensor";

export class ImageLoader extends React.Component {
  static propTypes = {
    src: PropTypes.string.isRequired,
    transitionTime: PropTypes.string,
    lazyLoad: PropTypes.bool,
    children: PropTypes.oneOfType([
      PropTypes.arrayOf(PropTypes.node),
      PropTypes.node,
      PropTypes.func,
    ]),
  };

  state = {
    hasLoaded: false,
    hasFailed: false,
    shouldShowLoader: true,
    visibilitySensorIsActive: true,
  };

  componentDidMount() {
    if (!this.props.lazyLoad) {
      this.preloadImage();
    }
  }

  componentDidUpdate(prevProps) {
    if (prevProps.src !== this.props.src) {
      if (!this.props.lazyLoad) {
        this.preloadImage();
      } else {
        this.setState({ visibilitySensorIsActive: true });
      }
    }
  }

  render() {
    const { src, children, lazyLoad } = this.props;
    const {
      hasLoaded,
      hasFailed,
      shouldShowLoader,
      visibilitySensorIsActive,
    } = this.state;
    /*
     * When using ImageLoader the children prop should be a function.
     * Render calls this function with the props it requires to
     * fade in an image.
     * See https://reactjs.org/docs/render-props.html
     */
    const childElement = children({
      hasLoaded,
      shouldShowLoader,
      hasFailed,
      src,
    });

    /*
     * Wrap element in react visibility sensor when lazyLoad is true
     */
    return lazyLoad ? (
      <VisibilitySensor
        active={visibilitySensorIsActive}
        onChange={this.handleVisibilityChange}
        partialVisibility
      >
        {childElement}
      </VisibilitySensor>
    ) : (
      childElement
    );
  }

  preloadImage() {
    const { src } = this.props;
    preloader
      .simplePreload(src)
      .then(this.handleImageLoaderLoaded)
      .catch(this.handleImageLoaderFailed);
  }

  /*
   * Once the image is loaded successfully this handler will be
   * called, this will change the `hasLoaded` attribute to true
   * then the `shouldShowLoader` attribute to false after
   * the transition has completed. The reason we do this on a
   * timeout is to ensure the loader is not hidden as soon as the
   * image has loaded resulting in a 'flash'
   */
  handleImageLoaderLoaded = () => {
    const { transitionTime } = this.props;
    const hideLoaderDelay = cssTimeToMs(transitionTime);
    this.setState({ hasLoaded: true }, () => {
      setTimeout(() => {
        this.setState({
          shouldShowLoader: false,
        });
      }, hideLoaderDelay);
    });
  };

  handleImageLoaderFailed = () => {
    this.setState({ hasFailed: true });
  };

  handleVisibilityChange = isVisible => {
    if (!isVisible) return;
    this.setState({ visibilitySensorIsActive: false });
    this.preloadImage();
  };
}

ImageLoader.defaultProps = {
  transitionTime: defaults.transitionTime,
};

export default ImageLoader;
