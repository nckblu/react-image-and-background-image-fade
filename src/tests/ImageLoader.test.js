import React from "react";
import ImageLoader from "../components/ImageLoader";
import renderer from "react-test-renderer";
import { shallow } from "enzyme";
import { flushPromises } from "./util";
import { FAIL_SRC, simplePreloadMock } from "./__mocks__/ImageLoader";

jest.mock("image-preloader");
const ImagePreloader = require("image-preloader");
ImagePreloader.simplePreload.mockImplementation(simplePreloadMock);

describe("ImageLoader", () => {
  beforeEach(() => {
    ImagePreloader.simplePreload.mockClear();
  });

  const genericProps = {
    src: "me-drinking-pina-colada.jpg",
    children: ({ src }) => <div className='oh-hey'>{src}</div>,
  };

  it("Matches snapshot of generic props", () => {
    const props = {
      ...genericProps,
    };
    const tree = renderer.create(<ImageLoader {...props} />).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it("Should call preload image with the src", () => {
    const props = { ...genericProps };
    shallow(<ImageLoader {...props} />);
    expect(ImagePreloader.simplePreload.mock.calls[0][0]).toEqual(props.src);
  });

  it("Should call preload image just the once", () => {
    const props = { ...genericProps };
    shallow(<ImageLoader {...props} />);
    expect(ImagePreloader.simplePreload.mock.calls.length).toEqual(1);
  });

  it("Should set hasLoaded to true once the image has loaded", async () => {
    const props = { ...genericProps };
    const wrapper = shallow(<ImageLoader {...props} />);
    await flushPromises();
    wrapper.update();
    expect(wrapper.state("hasLoaded")).toEqual(true);
  });

  it("Should keep 'shouldShowLoader' true for the specified transitionTime", async () => {
    jest.useFakeTimers();
    const props = { ...genericProps, transitionTime: "0.5s" };
    const wrapper = shallow(<ImageLoader {...props} />);
    await flushPromises();
    wrapper.update();
    expect(wrapper.state("shouldShowLoader")).toEqual(true);
    jest.runTimersToTime(500);
    expect(wrapper.state("shouldShowLoader")).toEqual(false);
    jest.useRealTimers();
  });

  it("Reacts properly to a change in src prop", () => {
    const props = { ...genericProps };
    const newSrc = "pickle_rick.tiff";
    const wrapper = shallow(<ImageLoader {...props} />);
    expect(ImagePreloader.simplePreload.mock.calls[0][0]).toEqual(props.src);
    wrapper.setProps({ src: newSrc });
    expect(ImagePreloader.simplePreload.mock.calls[1][0]).toEqual(newSrc);
  });

  it("Should set hasFailed to true when there is a failure to load image", async () => {
    const props = { ...genericProps, src: FAIL_SRC };
    const wrapper = shallow(<ImageLoader {...props} />);
    await flushPromises();
    wrapper.update();
    expect(wrapper.state("hasFailed")).toEqual(true);
  });
});
