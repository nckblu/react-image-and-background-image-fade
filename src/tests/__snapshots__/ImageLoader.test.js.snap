// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ImageLoader Matches snapshot of generic props 1`] = `
<div
  className="oh-hey"
>
  me-drinking-pina-colada.jpg
</div>
`;
